SAMZA-1178: Generate JSON from StreamPlan

xinyuiscool opened the pull request:
As the first step to visualize the StreamGraph/Plan, this patch generates a json representation of it. For the example StreamGraph in TestPlanJsonGenerator, here is the json that is generated (https://github.com/xinyuiscool/examples/blob/master/example_plan.json). This patch also introduces the immutable StreamPlan and did some code cleanup.

xinyuiscool commented:
@jmakes and @prateekm : please take a look at this patch. Thanks.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/JobGraph.java:
```diff
@@ -45,7 +45,7 @@
  * So the graph may have cycles and it's not a DAG.
  */
-public class JobGraph {
+/* package private */ class JobGraph {
```
Minor: Can we just leave out the `/*package private */` everywhere?

xinyuiscool replied:
Good catch! fixed.

jmakes replied:
fwiw, I like it. It assures future readers that the qualifier wasn't forgotten and that it was package private by design.

prateekm replied:
@jmakes This seems pretty reasonable. It was on every method in this class earlier.

prateekm commented on samza-core/src/main/java/org/apache/samza/runtime/RemoteApplicationRunner.java:
```diff
@@ -102,12 +94,11 @@ public ApplicationStatus status(StreamApplication app) {
         JobRunner runner = new JobRunner(jobConfig);
         ApplicationStatus status = runner.status();
-        log.debug("Status is {} for jopb {}", new Object[]{status, job.getId()});
+        log.debug("Status is {} for jopb {}", new Object[]{status, jobConfig.getName()});
```
Typo in `jobp`.

xinyuiscool replied:
Fixed.

prateekm commented on samza-core/src/test/java/org/apache/samza/execution/TestUtil.java:
```diff
@@ -0,0 +1,117 @@
+import org.apache.samza.system.SystemStreamMetadata;
+import org.apache.samza.system.SystemStreamPartition;
+
+public class TestUtil {
```
Prefer more descriptive/specific name.

xinyuiscool replied:
Fixed.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java:
```diff
@@ -57,7 +59,7 @@ public ExecutionPlanner(Config config, StreamManager streamManager) {
   }
 
-  public JobGraph plan(StreamGraph streamGraph) throws Exception {
+  public StreamPlan plan(StreamGraph streamGraph) throws Exception {
```
`StreamPlan` doesn't seem very meaningful. Maybe `ExecutionPlan`?

xinyuiscool replied:
I name it StreamPlan to be consistent with StreamApplication, StreamGraph, StreakTask .. :). In Flink it's called StreamingPlan so I guess I will keep it for now.

prateekm replied:
StreamingPlan seems fine, but StreamPlan doesn't make sense by itself. @jmakes, any thoughts? We can also rename later, not a blocker.

jmakes replied:
Putting it another way, I think it's more accurate to say this object is "a plan for execution" than "a plan for streaming". Hence ExecutionPlan.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java:
```diff
@@ -66,7 +68,12 @@ public JobGraph plan(StreamGraph streamGraph) throws Exception {
+    List<StreamSpec> intermediateStreams = jobGraph.getIntermediateStreams().stream()
+        .map(streamEdge -> streamEdge.getStreamSpec())
+        .collect(Collectors.toList());
+    return new StreamPlan(streamGraph, jobConfigs, intermediateStreams, planJson);
```
It's a little weird that the JSON representation of the plan is required to construct it, and that this class needs to know about the PlanJsonGenerator. Maybe add a helper to StreamPlan that returns the JSON representation instead? Maybe even override its toString?

xinyuiscool replied:
Fixed. Now the plan has both StreamGraph and JobGraph, and the objects/methods are internal to the plan.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java:
```diff
@@ -0,0 +1,266 @@
+    }
+  }
+
+  public static final class GraphNodes {
```
It's a little strange that there are public classes for all these nodes that are only there to generate the Plan representation.

xinyuiscool replied:
Oh, I think the public is not needed here. I removed it to be package private (for testing).

xinyuiscool commented:
@prateekm : I think I also fixed the stuff you mentioned in your early feedback. As we discussed, NodeId is the ID for either operator node or stream node, and it is used to construct the topology of the graph. For rendering, stream is also a node in this json graph (we will start from the source stream). For join, now two operator nodes are created to keep it in sync with StreamGraph. Each node has a reference of the other in the pair. I updated the link to the example json. Please take a look again.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java:
```diff
@@ -0,0 +1,295 @@
+
+  private static abstract class GraphNode {
+    private Integer nodeId;
+    private Set<Integer> successors = new HashSet<>();
```
Maybe `nextNodes` to indicate that the int is the nodeId (as opposed to streamId/opId)?

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java:
```diff
@@ -0,0 +1,295 @@
+  static final class OpNode extends GraphNode {
+    private String opCode;
+    private int opId;
+    private int pairNodeId = -1;  //for join operator, we will have a pair nodes for two partial joins
```
Naming: `pairedNodeId`

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java:
```diff
@@ -0,0 +1,295 @@
+    return node;
+  }
+
+  private static <T extends GraphNode> List<T> createListAndSort(Collection<T> nodes) {
```
Minor: Name makes it sound like a more general helper than it is. Maybe `sortGraphNodesById`?

xinyuiscool replied:
Removed. Did a lot of clean up and now it should be much better to read now.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java:
```diff
@@ -0,0 +1,295 @@
+      node.getSuccessors().add(next.getNodeId());
+
+      if (spec instanceof SinkOperatorSpec) {
+        StreamEdge streamEdge = outputToStreamSpecs.get(((SinkOperatorSpec) spec).getOutStream());
```
How does this work? After my changes in https://github.com/apache/samza/pull/92, SinkOperatorSpec#getNextStream returns null since it's a terminal operator. Did I get something wrong?

jmakes replied:
@prateekm getNextStream() returns null but getOutStream() doesn't. It seems a Sink has an OutputStream but not an output MessageStream

prateekm replied:
getOutStream() was renamed to getNextStream() in #92 to distinguish b/w actual/physical output streams and the next MessageStream in the operator DAG. The OutputStream in SinkOperatorSpec is the one that actual output is sent to. It has no next stream since it's a terminal operator. partitionBy returns a new InputStream to chain further operations on. Does that make sense?

jmakes replied:
Is it committed? https://github.com/apache/samza/blob/8515448a2023ae6c78b9b0bb8e297cf346775e13/samza-core/src/main/java/org/apache/samza/operators/spec/SinkOperatorSpec.java#L109

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java:
```diff
@@ -0,0 +1,295 @@
+  // (who register first in the map wins).
+  Map<MessageStream, OperatorSpec> outputStreamToJoinSpec = new HashMap<>();
+
+  /* package private */ String toJson(JobGraph jobGraph) throws Exception {
```
Looks like this class is stateful and is meant to be created once and saved for reuse. Is this method idempotent?

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java:
```diff
@@ -0,0 +1,295 @@
+    return new String(out.toByteArray());
+  }
+
+  private Job buildJob(JobNode jobNode, StreamGraphImpl streamGraph, JobGraph jobGraph) {
```
Minor: Maybe `toJob`?

jmakes replied:
There's definitely something odd about the naming or modularity. 

I'd expect a method called "buildJob"  to have an implementation that ONLY has the last 5 lines here.  So like:
`

  private Job buildJob(JobNode jobNode, Map<Integer, OpNode> opNodes) {

     Job job = new Job();
     job.setJobName(jobNode.getJobName());
     job.setJobId(jobNode.getJobId());
     job.setOpNodes(createListAndSort(opNodes.values()));
     return job;
   }

`
It does what its name implies. It builds a job and does NOTHING else.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java:
```diff
@@ -0,0 +1,295 @@
+    return job;
+  }
+
+  private void buildJsonGraph(MessageStream inputStream, GraphNode node, Map<Integer, OpNode> opNodes, Map<OutputStream, StreamEdge> outputToStreamSpecs) {
```
Needs documentation.

jmakes replied:
+1 Particularly because it seems to mutate some args and not others. It's not evident from the method signature what it does and how the output is captured.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java:
```diff
@@ -0,0 +1,295 @@
+ */
+public class PlanJsonGenerator {
+
+  private static abstract class GraphNode {
```
Should add documentation for these classes, and generally what the Graph structure & representation is.

jmakes commented on samza-core/src/main/java/org/apache/samza/execution/StreamPlan.java:
```diff
@@ -0,0 +1,77 @@
+ */
+public class StreamPlan {
+  private final StreamGraph streamGraph;
+  private final JobGraph jobGraph;
```
Commentary (no action in this PR):
This is the most compelling evidence yet that StreamGraph needs to be renamed. When I first read this, I thought the StreamPlan was keeping separate graphs to track the streams and the jobs. 

Now if they were named more meaningfully, like 
ApplicationLogic
JobGraph

I'd have a much better idea of what this plan was tracking.

xinyuiscool replied:
:). Based on the feedback, I refactored the code, and now ExecutionPlan is an interface, and JobModel implements it. I think now it looks cleaner.

jmakes commented on samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java:
```diff
@@ -0,0 +1,295 @@
+/**
+ * This class generates the json representation of the plan.
+ */
+public class PlanJsonGenerator {
```
This feels like an unnecessarily home-grown solution. 

Why not add an entry to SamzaObjectMapper implementing the Jackson JsonSerializer and let Jackson do the rest?

prateekm replied:
Agree that this feels odd, but prefer not adding to SamzaObjectMapper. It seems like that class is named more generally than what it is - it's actually the JSON serdes for coordinator stream contents. Is that right?

jmakes replied:
@prateekm I don't see the stigma with SamzaObjectMapper. It could be renamed or split up if necessary.  As a compromise any other standard Jackson implementation would be fine. But we should not reinvent the wheel.

prateekm replied:
@jmakes To add context to what I meant: This class is also using the regular Jackson ObjectMapper to serialize its data (in lines 203-206). Rest/most of the code + logic is for creating the data structures/POJOs to be serialized. What' I'm proposing is that instead of creating new data structures for serialization it'd be better if we could serialize the existing classes (JobGraph, JobNode, StreamEdge). That would get rid of this `PlanJsonGenerator`. Whether the actual serialization is done using a global registry like SamzaObjectMapper or using an ObjectMapper directly would then be a separate discussion.

jmakes replied:
Makes sense and it sounds like that would help. My focus was on the apparent inversion of the logic to have this homegrown class as the encapsulation to convert the plan to json and using Jackson under the covers. That's opposed to exposing a standard Jackson JsonSerializer that has some custom logic under the covers.

jmakes commented on samza-core/src/test/java/org/apache/samza/execution/TestExecutionUtils.java:
```diff
@@ -0,0 +1,117 @@
+    };
+  }
+
+  static JoinFunction createJoin() {
```
nit: this really has nothing to do with "execution". Perhaps belongs in a separate TestOperatorUtils

prateekm replied:
Also nothing to do with Operators, it's just some common code b/w ExecutionPlanner layer (and below) test classes. One of these methods (createRunner) will probably be irrelevant after the refactor in #92. Also, @xinyuiscool why not use mocks for these instead of creating actual instances?

xinyuiscool replied:
Good point! I used mock for JoinFunction. ApplicationRunner also uses mock from @prateekm 's patch. The class is removed now.

xinyuiscool commented:
I merged @prateekm 's recent patch and addressed all the comments. Please take a look again. Thanks!

xinyuiscool commented:
btw, also updated the example json here (https://github.com/xinyuiscool/examples/blob/master/example_plan.json).

jmakes commented on samza-core/src/main/java/org/apache/samza/runtime/RemoteApplicationRunner.java:
```diff
@@ -59,18 +56,14 @@ public RemoteApplicationRunner(Config config) {
     try {
       // 1. initialize and plan
-      JobGraph jobGraph = getExecutionPlan(app);
+      ExecutionPlan plan = getExecutionPlan(app);
```
This whole class reads much better now. Nice work!

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/JobGraph.java:
```diff
@@ -54,23 +55,49 @@
+   * Returns the configs for single stage job, in the order of topologically sort.
+   * @return list of job configs
+   */
+  public List<JobConfig> getJobConfigs() {
```
Minor: @Override for public interface methods and remove javadoc (it'll be inherited).

prateekm commented on samza-core/src/test/java/org/apache/samza/execution/TestExecutionPlanner.java:
```diff
@@ -65,26 +65,7 @@
-  private SystemAdmin createSystemAdmin(Map<String, Integer> streamToPartitions) {
+  static SystemAdmin createSystemAdmin(Map<String, Integer> streamToPartitions) {
 
     return new SystemAdmin() {
```
Would still prefer using mocks. Makes the expectations in tests much more clearer.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java:
```diff
@@ -0,0 +1,233 @@
+/**
+ * This class generates the JSON representation of the {@link JobGraph}.
+ */
+public class JobGraphJsonGenerator {
```
Minor: package private?

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java:
```diff
@@ -0,0 +1,233 @@
+   * This class provides the necessary connection of operators for traversal.
+   */
+  static abstract class Traversable {
+    @JsonProperty("NextOperatorIds")
```
Let's use camelCase for field names in JSON.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java:
```diff
@@ -0,0 +1,233 @@
+
+  static final class StreamSpecJson {
+    @JsonProperty("Id")
+    String id;
```
`streamId`

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  static final class OperatorGraphJson {
+    @JsonProperty("InputStreams")
```
OutputStreams should be here as well for consistency.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  static final class StreamEdgeJson {
+    @JsonProperty("StreamSpec")
```
Let's add source and destination jobIds as well to represent the connectivity information.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  /**
+   * Traverse the {@StreamGraph} recursively and update the operator graph JSON POJO.
```
Minor: Will be good to add more descriptive documentation (and maybe use a more specific name) about what this is doing.

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java:
```diff
@@ -0,0 +1,244 @@
+   * @param opGraph {@link org.apache.samza.execution.JobGraphJsonGenerator.OperatorGraphJson}
+   * @return {@link org.apache.samza.execution.JobGraphJsonGenerator.OperatorJson}
+   */
+  private OperatorJson getOrCreateOperatorJson(OperatorSpec opSpec, OperatorGraphJson opGraph) {
```
Why getOrCreate? Should probably be just buildOperatorJson?

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java:
```diff
@@ -0,0 +1,244 @@
+   * @param streamEdges map of streamId to {@link org.apache.samza.execution.JobGraphJsonGenerator.StreamEdgeJson}
+   * @return {@link org.apache.samza.execution.JobGraphJsonGenerator.StreamEdgeJson}
+   */
+  private StreamEdgeJson getOrCreateStreamEdgeJson(StreamEdge edge, Map<String, StreamEdgeJson> streamEdges) {
```
Same as previous, should probably be just buildStreamEdgeJson?

prateekm commented on samza-core/src/main/java/org/apache/samza/execution/ExecutionPlan.java:
```diff
@@ -0,0 +1,51 @@
+  /**
+   * Returns the JSON representation of the plan for visualization
+   * @return json string
+   * @throws Exception exception
```
Can remove "for visualization". E.g., could be used for logging.
"@return json string representation of the plan"
Document what exception.
